Deprecated messages about passing null in widgets.php

snehapatil2001 opened the pull request:
## Ticket
https://core.trac.wordpress.org/ticket/61179

## Description
- This **PR** addresses a deprecation warning in the `widgets.php` file when running WordPress on PHP 8.3. The warning occurs because `html_entity_decode()` is called with `null` as its first parameter, which is deprecated in PHP 8.3.

## Changes
- Updated the `html_entity_decode` call in `widgets.php` to use the null coalescing operator (`??`) to provide an empty string as a default value if `$item->get_description()` returns `null`.

## Testing Instructions

1. **Automated Testing**:
   - Added unit tests for handling null, empty string, and valid string descriptions.
   - Run the tests with PHPUnit to ensure the changes work as expected.

2. **Manual Testing**:
   - **Null Description**: Modify a widget to return `null` for its description. Check that no deprecation warnings are logged and the output is correct.
   - **Empty String Description**: Modify a widget to return an empty string for its description. Verify that no unnecessary output is rendered.
   - **Valid String Description**: Use a widget with a valid description and ensure it displays correctly on the frontend.

snehapatil2001 committed:
fix: null coalescing operator is used to ensure a string is always
passed to the function.

snehapatil2001 committed:
fix: unit tests.

mukeshpanchal27 commented on src/wp-includes/widgets.php:
```diff
@@ -1635,7 +1635,7 @@ function wp_widget_rss_output( $rss, $args = array() ) {
 		}
 
-		$desc = html_entity_decode( $item->get_description(), ENT_QUOTES, get_option( 'blog_charset' ) );
+		$desc = html_entity_decode( $item->get_description() ?? '', ENT_QUOTES, get_option( 'blog_charset' ) );
```
Instead if checking empty string can we check above if description is not empty then check other functions?

snehapatil2001 committed:
fix: check if RSS item description is not empty before processing.

snehapatil2001 committed:
Merge branch 'trunk' into widget-depreacation-warning

snehapatil2001 commented:
@mukeshpanchal27

Changes made:
- Modified the RSS widget output function in wp-includes/widgets.php
- Added a check to ensure the RSS item description is not empty before processing
- If the description is empty, we set $desc to an empty string
- This prevents calling html_entity_decode() with a null value

mukeshpanchal27 commented on src/wp-includes/widgets.php:
```diff
@@ -1635,8 +1635,13 @@ function wp_widget_rss_output( $rss, $args = array() ) {
-		$desc = html_entity_decode( $item->get_description(), ENT_QUOTES, get_option( 'blog_charset' ) );
-		$desc = esc_attr( wp_trim_words( $desc, 55, ' [&hellip;]' ) );
+		$description = $item->get_description();
+		if ( ! empty( $description ) ) {
+			$desc = html_entity_decode( $description, ENT_QUOTES, get_option( 'blog_charset' ) );
+			$desc = esc_attr( wp_trim_words( $desc, 55, ' [&hellip;]' ) );
+		} else {
+			$desc = '';
+		}
```
```suggestion
		$description = $item->get_description();
		$desc        = '';
		if ( ! empty( $description ) ) {
			$desc = html_entity_decode( $description, ENT_QUOTES, get_option( 'blog_charset' ) );
			$desc = esc_attr( wp_trim_words( $desc, 55, ' [&hellip;]' ) );
		}
```

mukeshpanchal27 commented on tests/phpunit/tests/widgets/wpWidgetDescription.php:
```diff
@@ -0,0 +1,50 @@
+	/**
+	 * Tests that html_entity_decode handles null description correctly.
+	 */
+	public function test_html_entity_decode_with_null_description() {
```
Add `@ticket` annotation for new tests

snehapatil2001 committed:
Update src/wp-includes/widgets.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

snehapatil2001 committed:
chore: add @ticket to function comments.

snehapatil2001 committed:
Merge branch 'trunk' into widget-depreacation-warning

snehapatil2001 commented:
@mukeshpanchal27  Done with the suggested changes.

snehapatil2001 committed:
Merge branch 'trunk' into widget-depreacation-warning

peterwilsoncc committed:
Merge branch 'trunk' into widget-depreacation-warning

peterwilsoncc reviewed (changes requested):
At the moment, these tests aren't quite right.

Rather than copying over the code written for the protection, they need to call `wp_widget_rss_output()` and ensure that the function does not produce an error. 

This ensures that the bug is picked up if `wp_widget_rss_output()` is rewritten in a way that re-introduces the bug. 

I've also noticed that the same issue can occur with the title and link if they are not provided within the item.

I think you'll be able to expand the tests in `tests/phpunit/tests/widgets/wpWidgetRss.php` to include test cases for malformed feeds to protect against deprecation warnings. The class uses a mock for valid feeds, it will need to include a mock file for invalid feeds too.
